### ISSUE TITLE
Multiple calls to HTML renderer does not update local state

### ISSUE THREAD

**Issue by kstewart83** [bikeshaving/crank#127]
So I was trying to use the crank HTML renderer in Node to understand the examples. I couldn't understand how to get examples with state working properly, and I can't find examples on the website where this is done. I ended up bringing in `jsdom` and using the dom renderer in order to get it to work. It seems counter-intuitive that the renderers would behave differently. Here's an example:

`index.tsx`
```tsx
import * as Crank from "@bikeshaving/crank";
import { renderer as dom } from "@bikeshaving/crank/cjs/dom";
import { renderer as html } from "@bikeshaving/crank/cjs/html";
import jsdomGlobal from "jsdom-global";

jsdomGlobal();

function *Counter() {
  let count = 0;
  while (true) {
    count++;
    yield (
      <div>
        You have updated this component {count} {count === 1 ? "time" : "times"}
      </div>
    );
  }
}

const c1 = <Counter />;
const c2 = <Counter />;

dom.render(c1, document.body);
console.log(document.body.innerHTML);
dom.render(c1, document.body);
console.log(document.body.innerHTML);

console.log("\nbreak\n");

console.log(html.render(c2));
console.log(html.render(c2));
```

I have the following set in my tsconfig.json:

`tsconfig.json`
```json
    "jsx": "react",
    "jsxFactory": "Crank.createElement",
```

Running this produces:

```sh
<div>You have updated this component 1 time</div>
<div>You have updated this component 2 times</div>

break

<div>You have updated this component 1 time</div>
<div>You have updated this component 1 time</div>
```

I would have expected the second group to mirror the first. Am I misunderstanding how to use the HTML renderer, or is this a bug?

**Comment by brainkim** [comment]
@kstewart83 Thanks for checking out Crank! Typically, HTML renderers are stateless, but you can have stateful HTML renderers by passing in an object for the second value:

```jsx
const key = {};
console.log(html.render(c2, key));
console.log(html.render(c2, key));
```

The second parameter to render plays two roles. Firstly, for the DOM renderer, it is the parent which the produced DOM tree is rendered into, mirroring the behavior of most DOM libraries. Secondly, it is a key which stores previous renderings, so that calls to `renderer.render` remember the past result. Internally, the renderer stores results in a weakmap. So if you use an object and rerender with the same object, you can achieve the same statefulness with the html renderer.

I’m not really sure what the use-case for stateful HTML renderers are, and it’s not something I’ve fully tested or thought about, and the renderer API is in a little bit of flux right now, so I’m very curious if you have a more concrete use-case in mind!

**Comment by kstewart83** [comment]
Thanks, that worked!

I thought the design approach you laid out in your blog post was really appealing. I'm really interested to see where the project goes. I think I gravitated towards the HTML renderer simply because I didn't see the the need for the DOM in order to try and understand the basics of your approach. I'm just starting to use generators in my other projects, so I don't have a fully rounded out intuition for all the nuances. Fortunately getting jsdom working wasn't too difficult.

I've used React a little, but am by no means an expert and I'm currently using Stencil for my hobby projects. Although not perfect, I really appreciate the approach Stencil takes to pre-rendering versus something like SSR. I'm sure there are plenty of pros and cons for each approach, but my preferred architecture is to have the code all static pre-rendered and hosted on a CDN (I use AWS so S3+Cloudfront), and then have the app hydrate client side. Then I just host APIs serverlessly through AWS lambda with the database backed by RDS. Stencil is able to do quite a bit at "compile" time to pre-render preload links and the like. I know there is a another issue tracking hydration, but seemed to be geared towards maybe an SSR use case. I'd definitely be interested in a pre-render use case as well.

Keep up the great work!
